File: scripts/plot_fig4_4_spectrogram.py
"""图4-4：各模型降噪后时频谱图对比 (plasma配色)"""
import numpy as np
import soundfile as sf
import matplotlib.pyplot as plt
from pathlib import Path
from scipy.signal import spectrogram

from plot_utils import setup_style
setup_style()

project_root = Path(r'C:\Users\XUWEILUN\Desktop\HydroDenoise-TF')

clean_dir = project_root / 'data' / 'ShipsEar' / 'test1' / 'clean'
noisy_dir = project_root / 'data' / 'ShipsEar' / 'test1' / 'noisy'
crn_dir   = project_root / 'baselines' / 'CRN-causal' / 'data' / 'data' / 'datasets' / 'tt' / 'tt_test1'
ct_dir    = project_root / 'experiments' / 'conv_tasnet' / 'estimates' / 'tt_test1'
dp_dir    = project_root / 'experiments' / 'dprnn' / 'estimates' / 'tt_test1'
dc_dir    = project_root / 'experiments' / 'dcamf_net' / 'denoised' / 'ShipsEar_test1'

fs = 16000
nperseg, noverlap, nfft = 256, 200, 512
freq_lim = 4000

# ------- 选取与图4-2一致的样本 (瞬态最强段 DCAMF-Net SI-SNRi 最高) -------
def compute_sisnr(est, ref):
    est = est - est.mean(); ref = ref - ref.mean()
    dot = np.dot(est, ref)
    target = dot * ref / (np.dot(ref, ref) + 1e-8)
    noise = est - target
    return 10 * np.log10(np.sum(target**2) / (np.sum(noise**2) + 1e-8) + 1e-8)

clean_files = sorted(list(clean_dir.glob('*.wav')))
best_sii, best_idx = -np.inf, 0
win_len = int(0.05 * fs)

for k, f in enumerate(clean_files):
    dcamf_file = dc_dir / f'{k+1:06d}.wav'
    if not dcamf_file.exists(): continue
    clean, _ = sf.read(clean_dir / f.name); clean = clean.mean(axis=1) if clean.ndim > 1 else clean
    noisy, _ = sf.read(noisy_dir / f.name); noisy = noisy.mean(axis=1) if noisy.ndim > 1 else noisy
    dcamf, _ = sf.read(dcamf_file); dcamf = dcamf.mean(axis=1) if dcamf.ndim > 1 else dcamf
    L = min(len(clean), len(noisy), len(dcamf))
    clean, noisy, dcamf = clean[:L], noisy[:L], dcamf[:L]
    if L < win_len: continue
    energies = np.array([np.sum(noisy[s:s+win_len]**2) for s in range(0, L-win_len, win_len)])
    start = np.argmax(energies) * win_len
    end = min(start + win_len, L)
    si = compute_sisnr(dcamf[start:end], clean[start:end]) - compute_sisnr(noisy[start:end], clean[start:end])
    if si > best_sii: best_sii, best_idx = si, k

print(f'选定样本: {best_idx}')

# ------- 加载该样本的全部模型输出 -------
def load_est(est_dir, idx, est_type):
    if est_type == 1: f = est_dir / f'{idx}_sph_est.wav'
    elif est_type == 2: f = est_dir / f'{idx:06d}_sph_est.wav'
    else: f = est_dir / f'{idx+1:06d}.wav'
    s, _ = sf.read(f); return s.mean(axis=1) if s.ndim > 1 else s

fname = clean_files[best_idx].name
clean, _ = sf.read(clean_dir / fname); clean = clean.mean(axis=1) if clean.ndim > 1 else clean
noisy, _ = sf.read(noisy_dir / fname); noisy = noisy.mean(axis=1) if noisy.ndim > 1 else noisy
crn = load_est(crn_dir, best_idx, 1)
ct  = load_est(ct_dir, best_idx, 2)
dp  = load_est(dp_dir, best_idx, 2)
dc  = load_est(dc_dir, best_idx, 3)

min_len = min(len(clean), len(noisy), len(crn), len(ct), len(dp), len(dc))
clean, noisy = clean[:min_len], noisy[:min_len]
crn, ct, dp, dc = crn[:min_len], ct[:min_len], dp[:min_len], dc[:min_len]

# ------- 计算语谱图 -------
signals = [clean, noisy, crn, ct, dp, dc]
titles  = ['干净信号', '带噪信号', 'CRN', 'Conv-TasNet', 'DPRNN', 'DCAMF-Net']
all_psd = []

for sig in signals:
    f, t, Sxx = spectrogram(sig, fs, nperseg=nperseg, noverlap=noverlap, nfft=nfft)
    mask = f <= freq_lim
    all_psd.append(10 * np.log10(Sxx[mask] + 1e-10))

vmin = np.floor(min(np.concatenate([p.ravel() for p in all_psd])) / 5) * 5
vmax = np.ceil(max(np.concatenate([p.ravel() for p in all_psd])) / 5) * 5
f_plot = f[f <= freq_lim] / 1000

# ------- 绘图 -------
fig, axes = plt.subplots(3, 2, figsize=(14, 12))
for i, ax in enumerate(axes.flat):
    im = ax.pcolormesh(t, f_plot, all_psd[i], shading='auto', cmap='plasma',
                        vmin=vmin, vmax=vmax)
    ax.set_xlabel('时间 (s)')
    ax.set_ylabel('频率 (kHz)')
    ax.set_title(titles[i], fontweight='bold')
    fig.colorbar(im, ax=ax)

plt.subplots_adjust(hspace=0.45, wspace=0.25)

save_dir = project_root / 'figures'
save_dir.mkdir(exist_ok=True)
fig.savefig(save_dir / 'fig4-4_Spectrogram_Comparison.pdf', dpi=300, bbox_inches='tight')
fig.savefig(save_dir / 'fig4-4_Spectrogram_Comparison.png', dpi=300, bbox_inches='tight')
plt.show()
print(f'图4-4 已保存至 {save_dir}')

File: matlab/plot_fig4_4_spectrogram.m
%% 图4-4：各模型降噪后时频谱图对比
clear; clc; close all;

set(0, 'DefaultAxesFontName', 'SimHei');
set(0, 'DefaultTextFontName', 'SimHei');
set(0, 'DefaultAxesFontSize', 8);

% ==================== 路径 ====================
projectRoot = 'C:\Users\XUWEILUN\Desktop\HydroDenoise-TF';
cleanDir = fullfile(projectRoot, 'data', 'ShipsEar', 'test1', 'clean');
noisyDir = fullfile(projectRoot, 'data', 'ShipsEar', 'test1', 'noisy');
crnEstDir = fullfile(projectRoot, 'baselines', 'CRN-causal', 'data', 'data', 'datasets', 'tt', 'tt_test1');
convtasnetEstDir = fullfile(projectRoot, 'experiments', 'conv_tasnet', 'estimates', 'tt_test1');
dprnnEstDir = fullfile(projectRoot, 'experiments', 'dprnn', 'estimates', 'tt_test1');
dcamfEstDir = fullfile(projectRoot, 'experiments', 'dcamf_net', 'denoised', 'ShipsEar_test1');

% ==================== 参数 ====================
fs = 16000;
nfft_spec = 512;
win_spec = hamming(256);
noverlap_spec = 200;
freqLim = [0, 4000];

% ==================== 选取与图4-2一致的样本 ====================
fprintf('遍历测试集，选取与图4-2一致的样本...\n');
cleanFiles = dir(fullfile(cleanDir, '*.wav'));
nSamples = length(cleanFiles);
bestSIi = -inf; bestIdx = 1;
analysisWin = 0.05;

for k = 1:nSamples
    fname = cleanFiles(k).name;
    dcamfFile = fullfile(dcamfEstDir, sprintf('%06d.wav', k));
    if ~exist(dcamfFile, 'file'), continue; end

    [clean, ~] = audioread(fullfile(cleanDir, fname)); clean = mean(clean,2);
    [noisy, ~] = audioread(fullfile(noisyDir, fname)); noisy = mean(noisy,2);
    dcamf = mean(audioread(dcamfFile), 2);

    minLen = min([length(clean), length(noisy), length(dcamf)]);
    clean=clean(1:minLen); noisy=noisy(1:minLen); dcamf=dcamf(1:minLen);
    winLen = round(analysisWin*fs);
    if minLen < winLen, continue; end
    nWindows = minLen - winLen + 1;
    energies = zeros(nWindows, 1);
    for s = 1:winLen:nWindows
        energies(s) = sum(noisy(s:s+winLen-1).^2);
    end
    [~, maxIdx] = max(energies);
    startS = max(1, maxIdx);
    endS = min(minLen, startS+winLen-1);
    if endS-startS+1 < winLen, continue; end
    sisnr_in = compute_sisnr(noisy(startS:endS), clean(startS:endS));
    sisnr_out = compute_sisnr(dcamf(startS:endS), clean(startS:endS));
    if sisnr_out - sisnr_in > bestSIi
        bestSIi = sisnr_out - sisnr_in;
        bestIdx = k;
    end
end
fprintf('选定样本: %d\n', bestIdx);

% ==================== 加载音频 ====================
fname = cleanFiles(bestIdx).name;
[clean, ~] = audioread(fullfile(cleanDir, fname)); clean = mean(clean,2);
[noisy, ~] = audioread(fullfile(noisyDir, fname)); noisy = mean(noisy,2);

crn = load_est(crnEstDir, bestIdx, 1);
ct  = load_est(convtasnetEstDir, bestIdx, 2);
dp  = load_est(dprnnEstDir, bestIdx, 2);
dc  = load_est(dcamfEstDir, bestIdx, 3);

minLen = min([length(clean), length(noisy), length(crn), length(ct), length(dp), length(dc)]);
clean=clean(1:minLen); noisy=noisy(1:minLen);
crn=crn(1:minLen); ct=ct(1:minLen); dp=dp(1:minLen); dc=dc(1:minLen);

% ==================== 计算语谱图 ====================
[~, F, T, p_clean] = spectrogram(clean, win_spec, noverlap_spec, nfft_spec, fs);
[~, ~, ~, p_noisy] = spectrogram(noisy, win_spec, noverlap_spec, nfft_spec, fs);
[~, ~, ~, p_crn]   = spectrogram(crn,   win_spec, noverlap_spec, nfft_spec, fs);
[~, ~, ~, p_ct]    = spectrogram(ct,    win_spec, noverlap_spec, nfft_spec, fs);
[~, ~, ~, p_dp]    = spectrogram(dp,    win_spec, noverlap_spec, nfft_spec, fs);
[~, ~, ~, p_dc]    = spectrogram(dc,    win_spec, noverlap_spec, nfft_spec, fs);

idxF = (F >= freqLim(1)) & (F <= freqLim(2));
F_plot = F(idxF);

specs = {p_clean, p_noisy, p_crn, p_ct, p_dp, p_dc};
titles = {'干净信号', '带噪信号', 'CRN', 'Conv-TasNet', 'DPRNN', 'DCAMF-Net'};
allPxx = [];
for i = 1:6, allPxx = [allPxx; 10*log10(specs{i}(idxF,:)+1e-10)]; end
cMin = floor(min(allPxx(:))/5)*5;
cMax = ceil(max(allPxx(:))/5)*5;

% ==================== 绘图 ====================
figure('Units', 'centimeters', 'Position', [2, 2, 24, 18], 'Color', 'white');

for i = 1:6
    subplot(3, 2, i);
    imagesc(T, F_plot/1000, 10*log10(specs{i}(idxF,:)+1e-10));
    axis xy; colormap(jet); caxis([cMin, cMax]);
    xlabel('时间 (s)'); ylabel('频率 (kHz)');
    title(titles{i}, 'FontWeight', 'bold');
    colorbar('Location', 'eastoutside', 'FontSize', 7);
end

% ==================== 保存 ====================
saveDir = fullfile(projectRoot, 'figures');
if ~exist(saveDir, 'dir'), mkdir(saveDir); end
pdfPath = fullfile(saveDir, 'fig4-4_Spectrogram_Comparison.pdf');
pngPath = fullfile(saveDir, 'fig4-4_Spectrogram_Comparison.png');
exportgraphics(gcf, pdfPath, 'ContentType', 'vector');
saveas(gcf, pngPath);
fprintf('图4-4 语谱图对比已保存至 %s\n', saveDir);

% ==================== 函数 ====================
function si = compute_sisnr(est, ref)
    est = est(:) - mean(est(:));
    ref = ref(:) - mean(ref(:));
    dot_prod = sum(est .* ref);
    s_target = dot_prod * ref / (sum(ref.^2) + 1e-8);
    e_noise = est - s_target;
    si = 10 * log10(sum(s_target.^2) / (sum(e_noise.^2) + 1e-8) + 1e-8);
end

function est = load_est(estDir, idx, estType)
    if estType == 1
        f = fullfile(estDir, sprintf('%d_sph_est.wav', idx-1));
    elseif estType == 2
        f = fullfile(estDir, sprintf('%06d_sph_est.wav', idx-1));
    elseif estType == 3
        f = fullfile(estDir, sprintf('%06d.wav', idx));
    else, est = []; return;
    end
    if ~exist(f, 'file'), est = []; return; end
    est = mean(audioread(f), 2);
end
